src/lib/eddsa.c

```diff
@@ -0,0 +1,182 @@
+      goto end;
+
+   // First 32 bytesof key_bits are the EdDSA seed (private key)
+   // Second 32 bytes are the EdDSA public key
+
+   seckey->key.ecc.x = BN_bin2bn(key_bits, 32, NULL);
+   seckey->pubkey.key.ecc.oid_len = sizeof(ed25519_oid);
+   seckey->pubkey.key.ecc.oid = calloc(1, sizeof(ed25519_oid));
+   memcpy(seckey->pubkey.key.ecc.oid, ed25519_oid, sizeof(ed25519_oid));
+
+   // Hack to insert the required 0x40 prefix on the public key
+   key_bits[31] = 0x40;
+   seckey->pubkey.key.ecc.point = BN_bin2bn(key_bits + 31, 33, NULL);
+
+   retval = 1;
+
+   end:
+      botan_rng_destroy(rng);
+      botan_privkey_destroy(eddsa);
+      return retval;
+   }
+
+int pgp_eddsa_verify_hash(const BIGNUM* r,
+                          const BIGNUM* s,
+                          const uint8_t *         hash,
+                          size_t                  hash_len,
+                          const pgp_ecc_pubkey_t *pubkey)
+   {
+   botan_pubkey_t eddsa = NULL;
+   botan_pk_op_verify_t verify_op = NULL;
+   int result = 0;
+   uint8_t bn_buf[64];
+
+   // Check curve OID matches 25519
+   if (pubkey->oid_len != 9 || memcmp(pubkey->oid, ed25519_oid, 9) != 0)
+      goto done;
+
+   // Unexpected size for Ed25519 key
+   if (BN_num_bytes(pubkey->point) != 33)
+      goto done;
+
+   BN_bn2bin(pubkey->point, bn_buf);
+
+   // See draft-koch-eddsa-for-openpgp-04 section 3 "Point Format"
+   if(bn_buf[0] != 0x40)
+      goto done;
+
+   if (botan_pubkey_load_ed25519(&eddsa, bn_buf + 1))
+      goto done;
+
+   if (botan_pk_op_verify_create(&verify_op, eddsa, "Pure", 0) != 0)
+      goto done;
+
+   if (botan_pk_op_verify_update(verify_op, hash, hash_len) != 0)
+      goto done;
+
+   // Unexpected size for Ed25519 signature
+   if (BN_num_bytes(r) > 32 || BN_num_bytes(s) > 32)
+      goto done;
+
+   memset(bn_buf, 0, sizeof(bn_buf));
+   BN_bn2bin(r, bn_buf);
+   BN_bn2bin(s, bn_buf + 32);
+
+   result = (botan_pk_op_verify_finish(verify_op, bn_buf, 64) == 0) ? 1 : 0;
+
+   done:
+   botan_pk_op_verify_destroy(verify_op);
+   botan_pubkey_destroy(eddsa);
+   return result;
+   }
+
+int pgp_eddsa_sign_hash(BIGNUM* r,
+                        BIGNUM* s,
+                        const uint8_t *hash,
+                        size_t         hash_len,
+                        const pgp_ecc_seckey_t *seckey,
+                        const pgp_ecc_pubkey_t *pubkey)
+   {
+   botan_privkey_t eddsa = NULL;
+   botan_pk_op_sign_t sign_op = NULL;
+   botan_rng_t rng = NULL;
+   int result = -1;
+   uint8_t bn_buf[64];
+
+   // Check curve OID matches 25519
+   if (pubkey->oid_len != 9 || memcmp(pubkey->oid, ed25519_oid, 9) != 0)
+      {
+      goto done;
+      }
+
+   // Unexpected size for Ed25519 key
+   if (BN_num_bytes(seckey->x) > 32)
+      goto done;
+
+   if (botan_rng_init(&rng, NULL) != 0)
+      goto done;
+
+   BN_bn2bin(seckey->x, bn_buf);
+
+   if (botan_privkey_load_ed25519(&eddsa, bn_buf) != 0)
```

Comment:
Does it make sanse to call this method "botan_privkey_load_ec()" ? I think always be a scalar and can be used for any ECC priv key (and I can reuse it for ECDSA)